docs/developer-guide/custom-widgets/universal-widgets.md

```diff
@@ -0,0 +1,339 @@
+# Universal Widgets
+
+Widgets in deck.gl allow developers to create custom UI elements that are deeply integrated into the deck.gl rendering system. This guide covers the steps to implement widgets that are framework-agnostic, ensuring compatibility across various applications.
+
+## Implementing the Widget Lifecycle Functions
+
+The widget lifecycle functions define how a widget initializes, updates, and cleans up its integration with deck.gl.
+
+### Adding a Widget
+
+[`onAdd({deck, viewId}): HTMLElement?`](../../api-reference/core/widget.md#onadd) - This method provides deck.gl with the root DOM element of your widget. This element is positioned based on `placement` and `viewId` members.
+
+```ts
+import { type Widget } from '@deck.gl/core'
+
+class AwesomeWidget implements Widget {
+  onAdd({ deck, viewId }) {
+    const element = document.createElement('div');
+    // Initialize and style your element
+    return element
+  }
+}
+```
+
+### Updating Properties
+
+[`setProps(props)`](../../api-reference/core/widget.md#setprops) - This method is called whenever widget properties are updated. Use this to apply changes dynamically.
+
+```ts
+class AwesomeWidget implements Widget {
+  setProps(props: Partial<ZoomWidgetProps>) {
+    Object.assign(this.props, props);
+  }
+}
+```
+
+### Handling Viewport Changes
+
+[`onViewportChange(viewport: Viewport)`](../../api-reference/core/widget.md#onviewportchange) - Widgets can listen to viewport updates via the `onViewportChange` method. You can target specific viewports or listen to all. 
+
```

Comment:
- Maybe make it clear that the "targeting" of viewports is done by setting viewId. 
- Does "you" mean widget writer or widget user (suggestion is to write it out).

Reply:
Yes. I guess a list of viewIds would be a possible extension, right now it is one or all...